[Channe Chwa] iP

channne opened the pull request:
### DukePro
> "Your mind is for having ideas, not holding them." - David Allen ([source](https://dansilvestre.com/productivity-quotes))

DukePro frees your mind of having to remember things you need to do. It's,
- text-based
- easy to learn
- ~~FAST~~ _SUPER_ FAST to use

All you need to do is,
1. download it from [here](https://nus-cs2103-ay2122s2.github.io/website/admin/ip-w4.html).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you :wink:

And it is **FREE**!

Features:

- [x] Managing tasks
- [x] Managing deadlines
- [ ] Reminders (coming soon)

---

If you're a Java programmer, you can use it to practice Java too. Here's the `main` method:
```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

channne committed:
Level 1: Greet, Echo, Exit

channne committed:
Level 2

channne committed:
Level 3: Mark as Done

channne committed:
Level 4: ToDo, Event, Deadline

channne committed:
A-TextUiTesting: Automated Text UI Testing

channne committed:
A-TextUiTesting: Automated Text UI Testing

channne committed:
Level-5: Handle Errors

channne committed:
Level-6: Delete

channne committed:
Add Level-7 features

channne committed:
Add Level-8 features

channne committed:
Delete data directory

channne committed:
Add A-MoreOOP features

channne committed:
Add A-Packages features

channne committed:
Add A-JUnit features

channne committed:
Add A-JavaDoc features

channne committed:
Add A-CodingStandard features

channne committed:
Add Level-9 features

channne committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/duke/command/Duke.java

channne committed:
Merge commit 'd8398594b7bc43da5eb865321c5a50cec4b3923d' into branch-A-Gradle

channne committed:
Add A-CheckStyle modifications

channne committed:
Add Level-10 features

declanleeler commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,61 @@
+     *
+     * @throws InvalidActionException If already done.
+     */
+    public void markAsDone() throws InvalidActionException {
```
I like the naming of this boolean method as it follows the coding standard.

declanleeler commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,61 @@
+     *
+     * @throws InvalidActionException If already undone.
+     */
+    public void markUndone() throws InvalidActionException {
```
I like the naming of this boolean method as it follows the coding standard.

declanleeler commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,61 @@
+
+    protected Task(String name, boolean isDone) {
+        this.name = name;
+        this.isDone = isDone;
```
I like the naming of this boolean variable as it follows the coding standard.

declanleeler commented on src/main/java/duke/ui/Parser.java:
```diff
@@ -0,0 +1,207 @@
+            String[] inputs = s.split(" " + format + " ");
+            if (inputs.length < 2) {
+                throw new IncompleteInputException("D: D: D: The description of a " + command
+                        + " is incorrect or empty D: D: D:");
```
I like the 8 spaces for the indentation for the wrapped line. Great adherence to the coding standard.

declanleeler commented on src/main/java/duke/ui/Parser.java:
```diff
@@ -0,0 +1,207 @@
+        } else if (entrySplit[0].equals("D")) {
+            if (entrySplit[1].equals("1")) {
+                return new Deadline(entrySplit[2],
+                        LocalDate.parse(entrySplit[3], DateTimeFormatter.ofPattern("MMM dd yyyy")),
```
I like the 8 spaces for the indentation for the wrapped line. Great adherence to the coding standard.

declanleeler commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,72 @@
+     * @param isDone Whether the task is done yet.
+     * @throws DateTimeParseException If the given time is not in the YYYY-MM-DD format.
+     */
+    public Deadline(String name, String time, boolean isDone) throws DateTimeParseException {
```
I like the adherence to boolean variable naming conventions.

declanleeler reviewed (commented):
Very strong work with no coding violations that I noticed. A lot of effort can be seen to have been put in to adhere to the Java coding standard. In addition, the code was easy to read and follow.

zxgoh commented on src/main/java/duke/task/Todo.java:
```diff
@@ -0,0 +1,49 @@
+    }
+
+    /**
+     * Returns a String representation of the Todo to be entered in the storage file.
+     *
+     * @return A String representation of the Todo.
+     */
```
Perhaps for `@Override` javadocs, you might want to use [`@inheritDoc`](https://stackoverflow.com/questions/13327534/details-of-difference-between-see-and-inheritdoc) instead

channne replied:
Thanks for the suggestion!

zxgoh commented on src/test/java/duke/task/TaskTest.java:
```diff
@@ -0,0 +1,38 @@
+    }
+
+    @Test
+    public void markAsDone_completedTask_exceptionThrown() {
```
Method name seems lengthy, maybe shorten it but keeping it descriptive as well

zxgoh commented on src/main/java/duke/command/Duke.java:
```diff
@@ -0,0 +1,110 @@
+                    storage.addTask(latestTask);
+                    return generator.getAddTaskMessage(latestTask, taskList.size());
+                }
+            } catch (WrongInputException | IncompleteInputException e) {
+                return e.getMessage();
+            } catch (DateTimeParseException e) {
+                return "Error parsing date, please enter dates in YYYY-MM-DD format!";
+            } catch (IOException e) {
+                return generator.getIoErrorMessage();
+            } catch (Exception e) {
+                return e.getMessage();
```
Solid use of exception catching by handling each individual cases, avoiding the "pokemon exception handling"

zxgoh commented on src/main/java/duke/command/TaskList.java:
```diff
@@ -0,0 +1,108 @@
+    public ArrayList<Task> findItems(String description) {
+        ArrayList<Task> newList = new ArrayList<>();
+        for (Task item : list) {
+            String s = item.toString();
```
Instead of using `s`, perhaps use a more descriptive name such as:
- `searchedTask`
- `currentTask`
- `taskAtHand`

channne replied:
Good idea, thanks for pointing it out!

zxgoh reviewed (commented):
Overall, the code was well organized, with appropriate levels of abstraction.
Mostly, names were descriptive and comprehensive. No deep nesting was used, which is good. 👍 😃

channne committed:
Multiple classes: add assertions

Some methods rely on certain assumptions to work.

Explicitly stating these assumptions using assertions allows for easier
detection of bugs.

Let's add in assertions in places where important assumptions should
hold.

channne committed:
Multiple classes: refactor code to improve code quality

Certain methods contain poor coding practices such as uninformative
variable names.

Let's refactor the code such that it follows the code quality standards.

This makes it easier for future me/others to understand the code.

channne committed:
Merge pull request #1 from channne/branch-A-Assertions

Multiple classes: add assertions

channne committed:
Merge pull request #2 from channne/branch-A-CodeQuality

Refactor code

channne committed:
Create gradle.yml

channne committed:
Add base functionality for B-Reminders.

Currently Event and Deadline tasks have attached dates.

However, it is easy to forget these dates without reminders.

Let's add functionality to remind the user at a user-specified date
and time about the Events/Deadlines or even Todos if required.

A Reminder attribute has been added to Tasks to easily link Tasks to
Reminders. A method has also been added to the Parser to allow for
parsing Reminder commands.

This commit is the first part of multiple parts to implement the full
B-Reminders functionality. Users can now add Reminders to Tasks but the
functionality for actually reminding users once the reminder time is
reached has not been added.

channne committed:
Add additional validity check for reminder date/time specified

channne committed:
Add reminding functionality for B-Reminders

Reminders pop up as a dialog box from Duke.

channne committed:
Improve GUI and code quality

channne committed:
Add representative screenshot

channne committed:
Update README.md

channne committed:
Update README.md

channne committed:
Update README.md

channne committed:
Update README.md

channne committed:
Update README.md

channne committed:
Update README.md

channne committed:
Update README.md

channne committed:
Update README.md

channne committed:
Update README.md

channne committed:
Delete unused imports

channne committed:
Make minor edits to documentation

channne committed:
Update build file

channne committed:
Delete incorrect assertions

channne committed:
UG final edits

channne committed:
Update existing `Exception` classes

channne committed:
Update existing `Task` classes

channne committed:
Edit documentation

channne committed:
Refactor code to make it pretty

channne committed:
Minor formatting changes to UG
